storage/range.go

```diff
@@ -273,20 +276,80 @@ func (r *Range) IsFirstRange() bool {
 	return bytes.Equal(r.Desc().StartKey, engine.KeyMin)
 }
 
-// IsLeader returns true if this range replica is the raft leader.
-// TODO(spencer): this is always true for now.
-func (r *Range) IsLeader() bool {
-	return true
+// getLease returns the current leader lease.
+func (r *Range) getLease() *proto.Lease {
+	return (*proto.Lease)(atomic.LoadPointer(&r.lease))
+}
+
+// HasLeaderLease returns true if this range replica currently holds
+// the Raft leader lease. It's possible that the replica is not in
+// fact the leader, but holding the lease means that no other replica
+// may be the leader either. Leases may not overlap, though a gap
+// between successive lease holders is expected.
+func (r *Range) HasLeaderLease() bool {
+	if l := r.getLease(); l != nil &&
+		l.StoreID == r.rm.StoreID() && r.rm.Clock().PhysicalNow() < r.getLease().Expiration {
+		return true
+	}
+	return false
+}
+
+// grantLeaderLease sets the lease expiration for this range replica,
+// until which, subsequent calls to HasLeaderLease() will return
+// true. If this range replica is already the lease holder, the
+// expiration will be extended as indicated. Otherwise, all duties
+// required of the range leader are commenced, including gossipping
+// cluster and config info where appropriate and clearing the command
+// and timestamp caches.
+//
+// TODO(tobias): call this method when processing the lease command
+// in Range.executeCmd().
+func (r *Range) grantLeaderLease(lease *proto.Lease) {
+	// Set the new leader lease.
+	var oldTerm uint64
+	if l := r.getLease(); l != nil {
+		oldTerm = l.Term
+	}
+	if lease.Term >= oldTerm {
+		atomic.StorePointer(&r.lease, unsafe.Pointer(lease))
+	}
+
+	// If this replica holds the lease and it's being granted as part of
+	// a new term, as opposed to renewed before having expired, clear
+	// the timestamp and command caches and maybe start gossiping (if
+	// we're the first range).
+	if r.HasLeaderLease() && lease.Term > prevTerm {
+		r.Lock()
+		if r.closer != nil {
+			close(r.closer)
+		}
+		r.closer = make(chan struct{})
```

Comment:
Actually we probably shouldn't touch the closer at all and leave that to be a single-use closer for shutdown. The goroutines should individually check the leadership status every time they're about to do something, simply because we want them to stop on their own when the lease expires.
There could be an elegant use of `context.WithDeadline` here.